infra/src/main/kotlin/fi/fta/geoviite/infra/tracklayout/LayoutGeometry.kt

```diff
@@ -174,21 +196,25 @@ data class LayoutAlignment(
                 }
             } else {
                 val previous = segments[index - 1]
-                require(previous.points.last().isSame(segment.points.first(), LAYOUT_COORDINATE_DELTA)) {
-                    "Alignment segment doesn't start where the previous one ended: " + "alignment=$id segment=$index " + "length=${segment.length} prevLength=${previous.length} " + "diff=${
-                        lineLength(
-                            previous.points.last(), segment.points.first()
-                        )
-                    }"
+                require(previous.segmentEnd.isSame(segment.segmentStart, LAYOUT_COORDINATE_DELTA)) {
+                    "Alignment segment doesn't start where the previous one ended: " +
+                            "alignment=$id segment=$index length=${segment.length} prevLength=${previous.length} " +
+                            "diff=${lineLength(previous.segmentEnd, segment.segmentStart)}"
                 }
                 require(isSame(previous.startM + previous.length, segment.startM, LAYOUT_M_DELTA)) {
-                    "Alignment segment m-calculation should be continuous: " + "alignment=$id segment=$index prevEnd=${previous.startM + previous.length} start=${segment.startM}"
+                    "Alignment segment m-calculation should be continuous: " +
+                            "alignment=$id segment=$index " +
+                            "prevStart=${previous.startM} prevLength=${previous.length} nextStart=${segment.startM}"
                 }
             }
         }
     }
 
     fun withSegments(newSegments: List<LayoutSegment>) = copy(segments = newSegments)
+
+    fun takeFirst(count: Int): List<LayoutPoint> = segments.firstOrNull()?.takeFirst(count) ?: listOf()
+
+    fun takeLast(count: Int): List<LayoutPoint> = segments.lastOrNull()?.takeLast(count) ?: listOf()
```

Comment:
Pitäisiköhän näiden ottaa huomioon se, että `count` voi olla isompi luku kuin ensimmäisen/viimeisen segmentin pistemäärä? 🤔 Tällöin oletettavasti pitäisi ottaa loput seuraavalta/edelliseltä segmentiltä (ja sitä seuraavalta/edelliseltä)

Nykyisellään tämä ei liene mikään ongelma, mutta lähinnä jotain sopivaa tulevaisuutta varten tässä on vaara pieneen kompastukseen joka ei ole ilmeinen lukematta toteutusta.

Reply:
Joo tuo abstraktio taitaa vähän vuotaa. En taida jaksaa tehdä kikkailua jota ei kuitenkaan oikeasti mistään käytetä, joten lisäänpä vain tuonne segmentin takeFirst/-Last:iin assertion että segmentillä pitää olla vähintään niin monta pistettä. Sitten tuo räjähtää jos koitetaan käyttää laajemmin ja voidaan toteuttaa alignment-tason yhdistely tarpeen mukaan.